lib/sensors/GPS.h

```diff
@@ -0,0 +1,339 @@
+enum SentenceTypes
+{
+	NONE = 0,
+	NMEA,
+	UBX,
+	RTCM
+} currentSentence = NONE;
+
+class GPS {
+ public:
+	GPS(PinName tx, PinName rx, uint32_t baud_rate);
+	void configure(); //configures the GPS (note that serial loop must be started before this is run)
+	void saveConfiguration(); //saves configuration to persistent memory
+	double getLat(); //degrees
+	double getLong(); //degrees
+	double getAlt(); //meters above elipsoid
+	double getTime(); //seconds since midnight
+
+
+ private:
+	//configuration functions
+	void setPortOutput(uint8_t portID, uint8_t comSettings);
+	void setPortInput(uint8_t portID, uint8_t comSettings);
+	void getPortSettings(uint8_t portID);
+	void setSerialRate(uint32_t baudrate, uint8_t uartPort = COM_PORT_UART1);
+	void configureMessage(uint8_t msgClass, uint8_t msgID, uint8_t portID, uint8_t sendRate);
+	void setNavigationFrequency(uint8_t navFreq);
+
+	//data processing functions
+	void process(uint8_t incoming, ubxPacket *incomingUBX, uint8_t requestedClass, uint8_t requestedID);
+	void processUBX(uint8_t incoming, ubxPacket *incomingUBX, uint8_t requestedClass, uint8_t requestedID);
+	void processUBXpacket(ubxPacket *msg);
+	void calcChecksum(ubxPacket *msg);
+	void addToChecksum(uint8_t incoming); //Given an incoming byte, adjust rollingChecksumA/B
+
+	//serial functions
+	void sendSerialCommand(ubxPacket *outgoingUBX);
+	void onByteReceive();
+	bool waitForAck();
+
+	//bit manipulation functions
+	uint32_t extractLong(uint8_t spotToStart); //Combine four bytes from payload into long
+	uint16_t extractInt(uint8_t spotToStart);  //Combine two bytes from payload into int
+	uint8_t extractByte(uint8_t spotToStart);  //Get byte from payload
+
+	//payload arrays
+	uint8_t payloadAck[2]{}; // Holds the requested ACK/NACK
+	uint8_t payloadCfg[MAX_PAYLOAD_SIZE]{}; // Holds the requested data packet
+	uint8_t payloadBuf[2]{}; // Temporary buffer used to screen incoming packets or dump unrequested packets
+
+	//packet structures (for processing)
+	ubxPacket m_packetAck = {0, 0, 0, 0, 0, payloadAck, 0, 0, UBLOX_PACKET_VALIDITY_NOT_DEFINED, UBLOX_PACKET_VALIDITY_NOT_DEFINED};
+	ubxPacket m_packetCfg = {0, 0, 0, 0, 0, payloadCfg, 0, 0, UBLOX_PACKET_VALIDITY_NOT_DEFINED, UBLOX_PACKET_VALIDITY_NOT_DEFINED};
+	ubxPacket m_packetBuf = {0, 0, 0, 0, 0, payloadBuf, 0, 0, UBLOX_PACKET_VALIDITY_NOT_DEFINED, UBLOX_PACKET_VALIDITY_NOT_DEFINED};
+
+	//helper variables for processing functions
+	uint8_t m_rollingChecksumA; //Rolls forward as we receive incoming bytes. Checked against the last two A/B checksum bytes
+	uint8_t m_rollingChecksumB; //Rolls forward as we receive incoming bytes. Checked against the last two A/B checksum bytes
+	uint16_t m_ubxFrameCounter;
+	ublox_packet_buffer_e m_activePacketBuffer = UBLOX_PACKET_PACKETBUF;
+	bool m_ignoreThisPayload;
+
+	//navigation data struct
+	//TODO: check if each member needs to be volatile, or just the struct itself
```

Comment:
not sure why this struct itself needs to be volatile, all its fields are changed by code explicitly...

Reply:
I'm not sure if the volatile is necessary, but the fields are being populated in an ISR context (the processing chain is attached to the uart interface serial interrupt)

Reply:
my bad... didn't notice it's in an ISR